License Export (#2900)
modelica/ModelicaSpecification: chapters/annotations.tex

\end{example}

\section{License Texts}\label{license-texts}

> For a top-level class the {\lstinline!annotation(License="modelica:/ModelicaLibraryName/Resources/Licenses/MyLicense.txt")!}\annotationindex{License}, gives the license files for the class.

Author: Match the singular example by singular in the surrounding text:
```suggestion
For a top-level class the {\lstinline!annotation(License="modelica:/ModelicaLibraryName/Resources/Licenses/MyLicense.txt")!}\annotationindex{License}, gives a license file for the class.
```

The annotation may give a scalar value or an array of values.
The license files are human-readable files containing the license conditions.

> External functions may also contain this annotation (with similar contents), see \cref{annotations-for-external-libraries-and-include-files}.

Author: Better with _content_?
```suggestion
External functions may also contain this annotation (with similar content), see \cref{annotations-for-external-libraries-and-include-files}.
```

Reviewer: Yes, if we consider the meaning of the annotation - not the letters.

\end{example}

> \section{License Texts}\label{license-texts}

Reviewer 2: What do you think, could we also add a better example here (similar as in 18.9.2)?

Reviewer: We could, but I think it can wait.

For a top-level class the {\lstinline!annotation(License="modelica:/ModelicaLibraryName/Resources/Licenses/MyLicense.txt")!}\annotationindex{License}, gives a license file for the class.
The annotation may give a scalar value or an array of values.
The license files are human-readable files containing the license conditions.
If the annotation specifies an array of license files the conditions in all of them must be satisfied (i.e., it is not intended for dual-licensing).
External functions may also contain this annotation (with similar content), see \cref{annotations-for-external-libraries-and-include-files}.

> When generating a distributable version for a translated model the license files for used functions and top-level classes should be included.

Author: Is it a conscious decision to _not_ require that tools include the license files?

Reviewer: There are several possible reasons:
1. The vagueness of "distributable version" makes it odd to have a hard requirement. (Is an executable distributable if it can be copied or does it require more? Is it distributed if distributed internally in a company?)
2. A tool may have options to handle the license files in other ways; if you generate a distributable version for embedding in a device the license files should be delivered in a completely different way: verified by a legal department, and some of the license files might be skipped (after review).
3. If the library is internal to the ones translating the model it makes sense that they license owners can skip their own licenses.

In all cases one could just remove the license files afterwards, so people can anyway circumvent it.

Reviewer: Note that 'should' doesn't mean that it can be skipped at will, just that there _may_ exist valid reasons in particular circumstances to ignore a particular item, but the full implications must be understood and carefully weighed before choosing a different course.

Author: A nice way to resolve this would be to include a non-normative explanation of why a tool might choose to not include the license files.

Reviewer: -> @HansOlsson write explanation.

Author: @HansOlsson ping

Reviewer: Have done

Author: OK, let's continue at the added content.  Resolving.

\end{example}

\section{License Texts}\label{license-texts}

For a top-level class the {\lstinline!annotation(License="modelica:/ModelicaLibraryName/Resources/Licenses/MyLicense.txt")!}\annotationindex{License}, gives a license file for the class.
The annotation may give a scalar value or an array of values.
The license files are human-readable files containing the license conditions.

> If the annotation specifies an array of license files the conditions in all of them must be satisfied (i.e., it is not intended for dual-licensing).

Reviewer: The person(s) distributing the derivative work depending on them I would say.

When making a library with license files the one building/distributing that library should (of course) ensure that the licenses are valid for the library.

> If the annotation specifies an array of license files the conditions in all of them must be satisfied (i.e., it is not intended for dual-licensing).

Reviewer: I would say the general statement about this only being an aid for the user covers this as well.

External functions may also contain this annotation (with similar content), see \cref{annotations-for-external-libraries-and-include-files}.

> When generating a distributable version for a translated model the license files for used functions and top-level classes should be included.

Author: New paragraph to make exception and the final sentence more clearly referring to the _should_:
```suggestion

When generating a distributable version for a translated model the license files for used functions and top-level classes should be included.
```

> When generating a distributable version for a translated model the license files for used functions and top-level classes should be included.

Reviewer 3: Should "distributable translated model" be clarified even a little bit? Adding "e.g., an FMU" at least? We have read the ReadMe.md and know where it's coming from, but I am not sure it would be clear as it stands in the specification. Translation that is mentioned in the specification is explained as "expansion of inherited base classes; parameterization of base classes,..." (section 1.2).

Reviewer: Makes sense.

> Exceptions could be that licenses will be distributed in another way, either the same ones -- or there is an agreement to use different licenses (e.g., for internal libraries).

Author: I find this hard to parse and would prefer:
```suggestion
Exceptions could be that licenses will be distributed in another way, or that there is an agreement to include different licenses (e.g., for internal libraries).
```

> Note that the collection of license files is only intended as a help, thus there is no need for a strict definition of distributable version or stricter rules for including the license files.

Author: I'm trying to see for whom this help is… Am I getting it wrong if I believe this is what is meant?
```suggestion
Note that the \lstinline!License!-annotation only allows the library developer to assist the tool in doing what is legally required; in the end it is a user responsibility to ensure that required license files accompany distributable versions of models.
```

I'd like to skip the part about _strict definition_, since I personally would find it helpful if the specification made it a tool responsibility in at least some of the most common situations, so that I would have the specification on my side when complaining to my tool vendor until they implement this.

Reviewer: Turned into suggestion below.

\end{example}

\section{License Texts}\label{license-texts}

For a top-level class the {\lstinline!annotation(License="modelica:/ModelicaLibraryName/Resources/Licenses/MyLicense.txt")!}\annotationindex{License}, gives a license text file for the class.

> The annotation may give a scalar value or an array of values.

Author: Here is a concrete proposal for addressing @otronarp's https://github.com/modelica/ModelicaSpecification/pull/2900#issuecomment-2022343060:
```suggestion
The annotation may give a scalar string or an array of strings, where each string is a URI referencing a license text file, see \cref{external-resources}.
```

Reviewer: As I understand it this is the only minor issue to discuss after it was accepted.
We might either accept the PR without it and then have a separate PR for this proposal, or amend the PR (arguing that it is just a trivial change).

Reviewer 2: I also raised one other concern previously:

> it is inconsistent in one location presumablely

I belive it can be easily addressed now.

The license text files are human-readable files containing the license conditions.
If the annotation specifies an array of license text files the conditions in all of them must be satisfied (i.e., it is not intended for dual-licensing).
External functions may also contain this annotation (with similar content), see \cref{annotations-for-external-libraries-and-include-files}.

> When generating a distributable version for a translated model the license text files for used functions and top-level classes should be included.

Reviewer: ```suggestion
When generating a distributable version (e.g., an FMU) for a translated model the license text files for used functions and top-level classes should be included.
```
As suggested.

Reviewer: @eshmoylova - your comment above.

Reviewer 3: That would work for me.

Reviewer 2: Is FMU already declared/defined in the MoLang Spec?

Reviewer: @eshmoylova  still ok?

Reviewer 3: FMU is not defined or used so far. But we refer to C89 and C99, for example, without providing any explanation or links. So, it should be acceptable to refer to FMU like that as well. It might be nicer to make it a link.

Reviewer 2: Yes, that was the intention of my comment: FMU should be declared and not considered as given.

Author: If linked, I think it should be handled similar to gettext; as a citation with the link given in the bibliography.

Reviewer: We can add the link later, if needed, but is the current proposed text good enough for the moment?

Reviewer 3: Looks good.

Exceptions could be that licenses will be distributed in another way, or that there is an agreement to include different licenses (e.g., for internal libraries).

> Note that the collection of license text files is only intended as a help, thus there is no need for a strict definition of distributable version or stricter rules for including the license text files.

Reviewer: ```suggestion
Note that the \lstinline!License!-annotation only allows the library developer and tool to assist the user in doing what is legally required; in the end it is a user responsibility to ensure that required license files accompany distributable versions of models.
```
As suggested.

Reviewer: @henrikt-ma your comment above with changes.

Reviewer: @henrikt-ma are you ok with change based on your comment?

Author: Yes, it is in the same spirit.

> Note that the \lstinline!License!-annotation only allows the library developer and tool to assist the user in doing what is legally required; in the end it is a user responsibility to ensure that required license files accompany distributable versions of models.

Reviewer: ```suggestion
Note that the \lstinline!License!-annotation only allows the library developer and tool to assist the user in doing what is legally required; in the end it is a user responsibility to ensure that required license text files accompany distributable versions of models.
```
